### PR TITLE
DEVAUTH - All-local devenv sans cloud or third-party auth workflows

### PR DESCRIPTION
**Pull request by tptignor**
## Summary of Changes

 - Enables direct frontend/backend communication sans Login.gov/Cloud.gov
 - Drives off new REACT_APP_DEVAUTH env var; no material changes to code if unset
 - Testing based on new dev user
   - Install via web: ./manage.py generate_dev_user
   - Create your own dev users as needed

## How to Test
 - Start frontend and backend locally with REACT_APP_DEVAUTH set.
 - Create the dev user in a web container shell: ./manage.py generate_dev_user
 - Open http://localhost:3000/data-files in browser.
 - Submit a data file normally.
 - Check logs and DBs to verify expected behaviors.

### CODE REVIEW

**Comment by tptignor** [comment]
> ## [Codecov](https://app.codecov.io/gh/raft-tech/TANF-app/pull/2784?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=raft-tech) Report
> Attention: `15 lines` in your changes are missing coverage. Please review.
> 
> > Comparison is base [(`e5ccc2c`)](https://app.codecov.io/gh/raft-tech/TANF-app/commit/e5ccc2c0899915cc4c71800fd1c6779eba12dd3e?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=raft-tech) 92.80% compared to head [(`bdb42a8`)](https://app.codecov.io/gh/raft-tech/TANF-app/pull/2784?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=raft-tech) 93.08%.
> > Report is 5 commits behind head on develop.
> 
> Additional details and impacted files

Not sure what to make of this. Got this in an email that names three files. One (views.py) just has debug lines and the others aren't in the PR.

File: tdrs-frontend/src/index.js [hunk 1]
```diff
@@ -24,8 +24,28 @@ axios.defaults.xsrfCookieName = 'csrftoken'
...
+      username: devEmail,
+      token: 'local-cypress-token',
+    })
+    .then(function (response) {
+      console.log(response)
+    })
+    .catch(function (error) {
+      console.log(error)
+    })
+  store.dispatch({ type: 'SET_AUTH', payload: { devEmail } })
```
Comment (jtimpe) [inline tdrs-frontend/src/index.js:40]: i think the reason the profile page comes up blank is because the reducer expects a `user` object in the payload, something like

```js
{
  type: 'SET_AUTH'
  payload: {
    user: {
      first_name: 'string',
      last_name: 'string',
      stt: 12,
      roles: [
          'Developer'
      ]
    }
  }
}
```

the `fetchAuth` action function sets the auth state with the result of calling the `/auth_check` endpoint, i think this initial dispatch gets quickly overridden by that auth check, allowing the other pages to work. the profile page, however, immediately looks for the user's `roles` in state, which aren't set by this dispatch, and throws an exception. It's likely the cypress tests would have the same issue if they navigated to the profile page.

there's a couple of possible ways to address this 
- the bug in the profile page could be fixed, allowing the page to load and maybe replace the state with a `fetchAuth` call thereby covering up the initial dispatch payload
- change the dispatch payload to include the roles. this could be setting them blank and letting a `fetchAuth` take over like the above point, or by modifying what the `/cypress/login` endpoint returns to include the roles and put the entire response in state with this dispatch

Comment (tptignor) [reply tdrs-frontend/src/index.js:40]: Put some more time into this. A one char fix neatly resolves the bug, allowing initialization to complete. I am then seeing the expected auth with the Developer profile loaded quickly afterwards. Thanks for the pointer!

File: tdrs-backend/tdpservice/users/api/login.py [hunk 2]
```diff
@@ -395,6 +395,7 @@ class CypressLoginDotGovAuthenticationOverride(TokenAuthorizationOIDC):
 
     def post(self, request):
         """Create a session for the specified user, if they exist."""
+        logging.debug(f"{self.__class__.__name__}: {request} ; {request.data}")
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/users/api/login.py]: is the logging still needed or can it be removed?

File: tdrs-backend/tdpservice/users/authentication.py [hunk 3]
```diff
@@ -3,36 +3,49 @@
...
 logger = logging.getLogger(__name__)
 
 class CustomAuthentication(BaseAuthentication):
     """Define authentication and get user functions for custom authentication."""
 
     @staticmethod
-    def authenticate(username=None, login_gov_uuid=None, hhs_id=None):
+    def authenticate(request=None, username=None, login_gov_uuid=None, hhs_id=None):
         """Authenticate user with the request and username."""
+        # HACK: This method currently needs to support two unrelated workflows.
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/users/authentication.py]: is this workaround still needed with the use of the `/login/cypress` endpoint?

File: tdrs-frontend/src/index.js [hunk 4]
```diff
@@ -24,8 +24,28 @@ axios.defaults.xsrfCookieName = 'csrftoken'
...
+  console.log(`dispatched SET_AUTH(${devEmail})`)
+}
+
 // call auth_check
 const store = configureStore()
+if (process.env.REACT_APP_DEVAUTH) {
```
Comment (jtimpe) [inline tdrs-frontend/src/index.js]: is it possible to add another condition here to prevent the `devLogin` process from happening if there's already an active session? i've noticed my non-devauth user sometimes getting replaced mid-session.

**Review by jtimpe — CHANGES_REQUESTED** [review]
i'd like to see the logging removed and perhaps another condition surrounding the `devLogin` function to prevent any session weirdness. 

the last question i'll just leave open for discussion with the comment that any changes to the login workflow or `CustomAuthentication` class that would affect production traffic, which this appears like it would, automatically give me pause. the implementation _looks_ okay to me (and i have not been able to break anything in my testing), but i don't feel confident that i could predict all the side effects of the change in other areas of the app. if that change can be undone or avoided, i'd feel a lot better. otherwise i'd just call on the team to pay a bit of extra attention to the `CustomAuthentication` class changes and test any edge cases we can think of.

File: tdrs-backend/tdpservice/users/api/login.py (hunk 2 above)
Comment (tptignor) [reply tdrs-backend/tdpservice/users/api/login.py]: Not required of course, but I found it helpful and thought it could still be useful at debug level; is it too much?

File: tdrs-backend/tdpservice/users/authentication.py (hunk 3 above)
Comment (tptignor) [reply tdrs-backend/tdpservice/users/authentication.py]: Yes, it is actually the meat of the fix. You can check the link for how the authenticate method is used by the rest framework.

**Comment by tptignor** [comment]
> i'd like to see the logging removed and perhaps another condition surrounding the `devLogin` function to prevent any session weirdness.
> 
> the last question i'll just leave open for discussion with the comment that any changes to the login workflow or `CustomAuthentication` class that would affect production traffic, which this appears like it would, automatically give me pause. the implementation _looks_ okay to me (and i have not been able to break anything in my testing), but i don't feel confident that i could predict all the side effects of the change in other areas of the app. if that change can be undone or avoided, i'd feel a lot better. otherwise i'd just call on the team to pay a bit of extra attention to the `CustomAuthentication` class changes and test any edge cases we can think of.

As mentioned in our recent meeting, that change is essential. There are two auth workflows which go through there. The implemented code enables us to support both without a larger redesign.

File: tdrs-frontend/src/index.js (hunk 4 above)
Comment (jtimpe) [reply tdrs-frontend/src/index.js]: to make the state selection easier, i'd suggest moving this logic to `App.js`, within the component. 

something like this, although the actual location to select in `auth = useSelector((state) => state.auth)` needs to be figured out to make the `!auth` condition work.
```js
  const dispatch = useDispatch()
  const auth = useSelector((state) => state.auth) // --> have to look in reducers for actual state path

  // call auth_check
  if (process.env.REACT_APP_DEVAUTH && !auth) {
    devLogin(dispatch, process.env.REACT_APP_DEVAUTH)
  }
  dispatch(fetchAuth())
```

the reason i'd like to move it is because the `App` component is already encapsulated by a redux `Provider` (in the render of `index.js`) meaning that `App` has access to the entire redux store through use of the `useDispatch` and `useSelector` hooks. Therefore you don't have to create a new store with `configureStore`

File: tdrs-backend/tdpservice/users/authentication.py [hunk 5]
```diff
@@ -3,36 +3,49 @@
...
-                return User.objects.get(login_gov_uuid=login_gov_uuid)
+                user_obj = User.objects.get(login_gov_uuid=login_gov_uuid)
             else:
-                return User.objects.get(username=username)
+                user_obj = User.objects.get(username=username)
         except User.DoesNotExist:
-            return None
+            user_obj = None
+        logging.debug(f"CustomAuthentication::authenticate found user: {user_obj}")
+        if type(request) == Request:
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/users/authentication.py]: Suggestion is to use Rest API authenticate functionality: https://github.com/encode/django-rest-framework/blob/047bec1288040b9f387e4977bb6cb3fd00d8ea40/rest_framework/authentication.py#L59C9-L59C21 which needs correct header and encoding.
The reason is then we can disable basic authentication specifically in prod as another layer of security.

File: tdrs-frontend/src/index.js (hunk 4 above)
Comment (tptignor) [reply tdrs-frontend/src/index.js]: Hi Jan,
I've been looking at this. There should be no problem with the !auth check. It should also be possible to move the code, but is that really an organizational improvement? Currently App.js isn't concerned with useDispatch or useSelector (and doesn't import them.) Also, the code implemented in the PR doesn't create a new store. The configureStore() invocation in the file is the original one (where the app initially creates the store.)

Comment (jtimpe) [reply tdrs-frontend/src/index.js]: you're right - i misread the diff, sorry about that. since you're building off a previous pattern, i'm ultimately good with this.

just to elaborate on my original point a little, though, i would argue the `App` component is the "right" place for the devauth logic (as well as the `fetchAuth` logic, probably) because in general i think it's a bad idea to have procedural code updates outside of a component. it's better if components are responsible for getting/setting state as it's more predictable when events will dispatch and when components will rerender, and i think it's more in-line with react/redux best practices
https://redux.js.org/style-guide/#use-the-react-redux-hooks-api
one area where i'd like to look more closely is the `getState()` call inside the conditional - when state is updated, a component utilizing `useSelector` will rerender automatically and any conditional logic will be re-performed. with `getState()` outside of the store-provided component, i'm not sure if this would normally have the intended effect.

because your implementation is working at the same time as the existing auth (reloading of `index.js` vs upon store update or component render), this is okay and it works. otherwise it may be prone to race conditions or other weird conflicting behavior. i think the frontend auth implementation could use a little work to make things clearer and more reliable, but your changes aren't negatively contributing to anything there.

**Comment by tptignor** [comment]
Sorry this is later than expected. I did some retesting in light of Jan's and Mo's comments. Jan was right. The feature now is enabled fully with the Cypress override workflow. With this design, the CustomAuthentication code isn't used to enable DEVAUTH. Moreover, in my testing, I found the previous changes to handle Django REST framework auth enabled GETs on the frontend /v1/auth_check to pass with just a username (and no cypress token.) In the last commits, you'll see that's all been reverted. The CustomAuthentication code has been highly distracting, so the only changes there now are 1. a comment regarding our discussions on future redesign/refactoring, 2. a small logging improvement which actually reduces log volume (and should be helpful when multiple requests come in at once.)

Please take a look when time allows, and if you're inspired, feel free to test it out.

**Comment by raftmsohani** [comment]
> Sorry this is later than expected. I did some retesting in light of Jan's and Mo's comments. Jan was right. The feature now is enabled fully with the Cypress override workflow. With this design, the CustomAuthentication code isn't used to enable DEVAUTH. Moreover, in my testing, I found the previous changes to handle Django REST framework auth enabled GETs on the frontend /v1/auth_check to pass with just a username (and no cypress token.) In the last commits, you'll see that's all been reverted. The CustomAuthentication code has been highly distracting, so the only changes there now are 1. a comment regarding our discussions on future redesign/refactoring, 2. a small logging improvement which actually reduces log volume (and should be helpful when multiple requests come in at once.)
> 
> Please take a look when time allows, and if you're inspired, feel free to test it out.

The fact that now the changes are minimal helps. I would just suggest a nit to leave a note for future security improvement. I still see a risk that someone will create this superuser with admin access in the backend and unfortunately even without the frontend the user is able to access /v1. A TODO suffices, but I don't want to leave this without a note. TODO will be to refactor the security to block the dev@test.com access to backend in prod

**Review by raftmsohani — APPROVED** [review]
LGTM!

**Review by jtimpe — APPROVED** [review]
Good work, thanks Tom!

**Review by elipe17 — APPROVED** [review]

**Comment by tptignor** [comment]
Thanks guys! I just merged. Mo, let's discuss any additional security concerns at the meeting tomorrow, and open a new ticket if we need to. Glad to see this committed, guys. Hopefully it helps future development.
